blog/2024-07-3-uppy-4.0.md

```diff
@@ -0,0 +1,364 @@
+services): `@uppy/aws-s3` and `@uppy/aws-s3-multpart`. They have different use
+cases. The advantages of multipart uploads are:
+
+- Improved throughput – You can upload parts in parallel to improve throughput.
+- Quick recovery from any network issues – Smaller part size minimizes the
+  impact of restarting a failed upload due to a network error.
+- Pause and resume object uploads – You can upload object parts over time. After
+  you initiate a multipart upload, there is no expiry; you must explicitly
+  complete or stop the multipart upload.
+- Begin an upload before you know the final object size – You can upload an
+  object as you are creating it.
+
+However, the downside is request overhead, as it needs to do creation, signing,
+and completion requests besides the upload requests. For example, if you are
+uploading files that are only a couple kilobytes with a 100ms roundtrip latency,
+you are spending 400ms on overhead and only a few milliseconds on uploading.
+This really adds up if you upload a lot of small files.
+
+AWS, and generally the internet from what we found, tend to agree that **you
+don’t want to use multipart uploads for files under 100MB**. But this sometimes
+puts users of our libraries in an awkward position, as their end users may not
+only upload very large files, or only small files. In this case a portion of
+their users get a subpar experience.
+
+---
+
+<!-- TODO: /aws-s3-multipart link should be /aws-s3 (needs site changes) -->
+
+We’ve merged the two plugins into `@uppy/aws-s3` with a new
+[`shouldUseMultipart`](/docs/aws-s3-multipart/#shouldusemultipartfile) option!
+By default it switches to multipart uploads if the file is larger than 100MB.
+You can pass a `boolean` or a function to determine it per file.
+
+## React hooks
+
+People working with React are more likely to create their own user interface on
+top of Uppy than those working with “vanilla” setups. Working with our pre-build
+UI components is a plug-and-play experience, but building on top of Uppy’s state
+with React primitives has been tedious.
+
+To address this we’re introducing to new hooks: `useUppyState` and
+`useUppyEvent`. Thanks to the TypeScript rewrite, we can now do powerful
+inference in hooks as well.
+
+### `useUppyState(uppy, selector)`
+
+Use this hook when you need to access Uppy’s state reactively.
+
+```js
+import Uppy from '@uppy/core';
+import { useUppyState } from '@uppy/react';
+
+// IMPORTANT: passing an initializer function to prevent Uppy from being recreated on every render.
+const [uppy] = useState(() => new Uppy());
+
+const files = useUppyState(uppy, (state) => state.files);
+const totalProgress = useUppyState(uppy, (state) => state.totalProgress);
+// We can also get specific plugin state.
+// Note that the value on `plugins` depends on the `id` of the plugin.
+const metaFields = useUppyState(
+	uppy,
+	(state) => state.plugins?.Dashboard?.metaFields,
+);
```

Comment:
Getting the following TS errors:

```
$ npx tsc --target esnext --module nodenext --noEmit index.ts                                  
node_modules/@uppy/dashboard/lib/Dashboard.d.ts:3:39 - error TS2307: Cannot find module '@uppy/core/lib/BasePlugin' or its corresponding type declarations.

3 import type { DefinePluginOpts } from '@uppy/core/lib/BasePlugin';
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/drag-drop/lib/DragDrop.d.ts:2:39 - error TS2307: Cannot find module '@uppy/core/lib/BasePlugin' or its corresponding type declarations.

2 import type { DefinePluginOpts } from '@uppy/core/lib/BasePlugin';
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/drag-drop/lib/DragDrop.d.ts:3:38 - error TS2307: Cannot find module '@uppy/core/lib/UIPlugin' or its corresponding type declarations.

3 import type { UIPluginOptions } from '@uppy/core/lib/UIPlugin';
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/progress-bar/lib/ProgressBar.d.ts:3:34 - error TS2307: Cannot find module '@uppy/core/lib/Uppy' or its corresponding type declarations.

3 import type { Uppy, State } from '@uppy/core/lib/Uppy';
                                   ~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/provider-views/lib/ProviderView/ProviderView.d.ts:4:36 - error TS2307: Cannot find module '@uppy/utils/lib/CompanionFile.ts' or its corresponding type declarations.

4 import type { CompanionFile } from '@uppy/utils/lib/CompanionFile.ts';
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/react/lib/Dashboard.d.ts:7:16 - error TS2833: Cannot find namespace 'React'. Did you mean 'preact'?

7 }, 'inline'> & React.BaseHTMLAttributes<HTMLDivElement>;
                 ~~~~~

node_modules/@uppy/react/lib/Dashboard.d.ts:19:51 - error TS2339: Property 'props' does not exist on type 'Dashboard<M, B>'.

19     componentDidUpdate(prevProps: Dashboard<M, B>['props']): void;
                                                     ~~~~~~~

node_modules/@uppy/react/lib/DashboardModal.d.ts:7:40 - error TS2833: Cannot find namespace 'React'. Did you mean 'preact'?

7 }, 'inline' | 'onRequestCloseModal'> & React.BaseHTMLAttributes<HTMLDivElement>;
                                         ~~~~~

node_modules/@uppy/react/lib/DashboardModal.d.ts:25:56 - error TS2339: Property 'props' does not exist on type 'DashboardModal<M, B>'.

25     componentDidUpdate(prevProps: DashboardModal<M, B>['props']): void;
                                                          ~~~~~~~

node_modules/@uppy/react/lib/DragDrop.d.ts:16:50 - error TS2339: Property 'props' does not exist on type 'DragDrop<M, B>'.

16     componentDidUpdate(prevProps: DragDrop<M, B>['props']): void;
                                                    ~~~~~~~

node_modules/@uppy/react/lib/ProgressBar.d.ts:15:53 - error TS2339: Property 'props' does not exist on type 'ProgressBar<M, B>'.

15     componentDidUpdate(prevProps: ProgressBar<M, B>['props']): void;
                                                       ~~~~~~~

node_modules/@uppy/react/lib/StatusBar.d.ts:16:51 - error TS2339: Property 'props' does not exist on type 'StatusBar<M, B>'.

16     componentDidUpdate(prevProps: StatusBar<M, B>['props']): void;
                                                     ~~~~~~~

node_modules/@uppy/status-bar/lib/StatusBar.d.ts:3:34 - error TS2307: Cannot find module '@uppy/core/lib/Uppy' or its corresponding type declarations.

3 import type { Uppy, State } from '@uppy/core/lib/Uppy';
                                   ~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/status-bar/lib/StatusBar.d.ts:4:39 - error TS2307: Cannot find module '@uppy/core/lib/BasePlugin' or its corresponding type declarations.

4 import type { DefinePluginOpts } from '@uppy/core/lib/BasePlugin';
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/status-bar/lib/StatusBarOptions.d.ts:1:38 - error TS2307: Cannot find module '@uppy/core/lib/UIPlugin' or its corresponding type declarations.

1 import type { UIPluginOptions } from '@uppy/core/lib/UIPlugin';
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~


Found 15 errors in 11 files.

Errors  Files
     1  node_modules/@uppy/dashboard/lib/Dashboard.d.ts:3
     2  node_modules/@uppy/drag-drop/lib/DragDrop.d.ts:2
     1  node_modules/@uppy/progress-bar/lib/ProgressBar.d.ts:3
     1  node_modules/@uppy/provider-views/lib/ProviderView/ProviderView.d.ts:4
     2  node_modules/@uppy/react/lib/Dashboard.d.ts:7
     2  node_modules/@uppy/react/lib/DashboardModal.d.ts:7
     1  node_modules/@uppy/react/lib/DragDrop.d.ts:16
     1  node_modules/@uppy/react/lib/ProgressBar.d.ts:15
     1  node_modules/@uppy/react/lib/StatusBar.d.ts:16
     2  node_modules/@uppy/status-bar/lib/StatusBar.d.ts:3
     1  node_modules/@uppy/status-bar/lib/StatusBarOptions.d.ts:1
```

Reply:
Probably better to test how this is actually used, so inside a component and with the right tsconfig.

https://stackblitz.com/edit/vitejs-vite-vehvbq?file=src%2FApp.tsx